### PR TITLE
add logic for handling copy_from argument

### LINKED ISSUE #131: Support creating a Quality Gate by copying existing Quality Gate
Currently the [`sonarqube_qualitygate`](https://registry.terraform.io/providers/jdamata/sonarqube/latest/docs/resources/sonarqube_qualitygate) resource only supports two inputs (`name` and `is_default`). 

It would be nice to be able to create Quality Gates by copying an existing gate, just as the `Copy` button in the UI (see image below), or the `api/qualitygates/copy` endpoint. 

bild

My proposed solution is to add an extra argument (`copy_from`) to the `sonarqube_qualitygate` resource to support copying an existing gate.

### PR DESCRIPTION
**Pull request by felixlut**
This PR introduces the `copy_from` argument when creating a SonarQube `Quality Gate`. 

It does so by calling the `api/qualitygates/copy` API as opposed to the regular `api/qualitygates/create` endpoint if the `copy_from` argument is supplied. 

Resolves #131

### CODE REVIEW

**Comment by felixlut** [comment]
Testing the copy functionality is not trivial since it's only really the conditions for it that are copied. This in combination that a Quality Gate is created with a number of conditions by default makes it a bit hard. 

My idea right now is to do the following steps:
1. Create a base Quality Gate (including the default conditons that comes with this call)
2. Create a custom condition for it on top of the default ones 
3. Create a copy of the base gate
4. Check that the custom condition is indeed in the list of conditions in the copy

Conceptually this is quite easy, but I'll probably dig into a bit in [custom validation functions](https://developer.hashicorp.com/terraform/plugin/testing/acceptance-tests/teststep#custom-check-functions) a bit to make it work

**Comment by felixlut** [comment]
> 

Ended up changing step 4 to chacking for equality between the `base` and `copied` gates

File: Makefile [hunk 1]
```diff
@@ -18,8 +18,8 @@ vet:
 	go vet ./...
 
 testacc:
-	docker run --name sonarqube1 -d -p 9001:9000 ${SONARQUBE_IMAGE}
```
Comment (jdamata) [inline Makefile:21]: can we revert these changes to the Makefile

File: local-terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+terraform {
```
Comment (jdamata) [inline local-terraform/main.tf:1]: if you'd like to keep this example of usage, can we move it into a new subdirectory under the examples directory: https://github.com/jdamata/terraform-provider-sonarqube/tree/master/examples

File: sonarqube/resource_sonarqube_qualitygate.go [hunk 3]
```diff
@@ -52,23 +61,63 @@ func resourceSonarqubeQualityGate() *schema.Resource {
...
-		"name": []string{d.Get("name").(string)},
-	}.Encode()
+
+	if gate_to_copy, ok := d.GetOk("copy_from"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/copy"
+		sonarQubeURL.RawQuery = url.Values{
+			"name":       []string{d.Get("name").(string)},
+			"sourceName": []string{gate_to_copy.(string)},
+		}.Encode()
+
```
Comment (jdamata) [inline sonarqube/resource_sonarqube_qualitygate.go:114]: please remove this whitespace

File: sonarqube/resource_sonarqube_qualitygate.go [hunk 4]
```diff
@@ -129,6 +178,12 @@ func resourceSonarqubeQualityGateRead(d *schema.ResourceData, m interface{}) err
 	d.Set("name", qualityGateReadResponse.Name)
 	// Api returns if true if set as default is available. when is_default=true setAsDefault=false so is_default=tue
 	d.Set("is_default", !qualityGateReadResponse.Actions.SetAsDefault)
+
```
Comment (jdamata) [inline sonarqube/resource_sonarqube_qualitygate.go:181]: Is `d.Set("conditions", qualityGateReadResponse.Conditions)` not suficient? 

I dont understand why we need to remarshal this back into json? If we really need this in json, can just use resp.Body? I think that should already be in json.

Comment (felixlut) [reply sonarqube/resource_sonarqube_qualitygate.go:180]: I thought so too, but It's not. Using `d.Set("conditions", qualityGateReadResponse.Conditions)` produces this error:
bild

I'm not the most adverse go-coder, so I can't explain exactly why, but using the below snippet works. 
```go
var allConditions []map[string]string
data, _ := json.Marshal(qualityGateReadResponse.Conditions)
json.Unmarshal(data, &allConditions)
d.Set("conditions", allConditions)
// d.Set("conditions", qualityGateReadResponse.Conditions)
```

Comment (felixlut) [reply sonarqube/resource_sonarqube_qualitygate.go:180]: I can dig into it more after work though

Comment (felixlut) [reply sonarqube/resource_sonarqube_qualitygate.go:180]: I based my solution for the schema on the Terraform provider for GitHub by the way, as seen [here](https://github.com/integrations/terraform-provider-github/blob/b4042f1c704d29f03f9619488a89e085b0b17b96/github/data_source_github_actions_organization_secrets.go#L15-L40)

**Review by jdamata — APPROVED** [review]

File: sonarqube/resource_sonarqube_qualitygate.go (hunk 4 above)
Comment (jdamata) [reply sonarqube/resource_sonarqube_qualitygate.go:181]: Thanks for the troubleshooting and in-depth explanation.
